MAINT: delay loading of backend_info to after imports

dschult opened the pull request:
Fixes #7671 

Delays updating backend_info and setting up configs until the end of the import process so that backends which import networkx inside their backend_info function don't run into a circular import issue.

We should test this locally with nx-cugraph installed just to make sure nothing strange occurs.

rossbar reviewed (approved):
This does indeed fix the runtime warning in the docs, thanks @dschult for chasing this down!

jarrodmillman reviewed (approved):
LGTM

dschult commented:
Thanks for verifying @rossbar . :)

I fixed a mypy issue and in the course of time-testing nx-parallel with this I caught a missing `nx.` in front of `config` that is now needed. greped for more missing `nx.` so I think that's OK now. I also moved the setting of doc_string for config.backends to inside the config setup function rather than in `__init__.py`.

I've checked it locally with nx-parallel installed as the only backend. And with no backends.
The CI checks are all passing including coverage.

dschult commented:
OK... This is passing now. @rlratzel has completed some other tests.  In it goes!
